ipaclient/install/ipa_client_samba.py

```diff
@@ -0,0 +1,726 @@
+        logger.info("Configured to use DNS discovery")
+
+    if api.env.host == server:
+        logger.error("Cannot run on IPA master")
+        print(
+            "Cannot configure Samba as a domain member on a domain "
+            "controller. Please use ipa-adtrust-install for that!"
+        )
+        return 1
+
+    if not options.netbiosname:
+        options.netbiosname = DNSName.from_text(api.env.host)[0].decode()
+    options.netbiosname = options.netbiosname.upper()
+
+    with use_api_as_principal(api.env.host_princ, paths.KRB5_KEYTAB):
+        try:
+            # Try to access 'service_add_smb' command, if it throws
+            # AttributeError exception, the IPA server doesn't support
+            # setting up Samba as a domain member.
+            service_add_smb = api.Command.service_add_smb
+
+            # Now try to see if SMB principal already exists
+            api.Command.service_show(api.env.smb_princ)
+
+            # If no exception was raised, the object exists.
+            # We cannot continue because we would break existing configuration
+            print(
+                "WARNING: SMB service principal %s already exists. "
+                "Please remove it before proceeding." % (api.env.smb_princ)
+            )
+            if not options.force:
+                return 1
+            # For --force, we should then delete cifs/.. service object
+            api.Command.service_del(api.env.smb_princ)
+        except AttributeError:
+            print(
+                "Chosen IPA master %s does not have support to"
+                "set up Samba domain members" % (server)
+            )
+            return 1
+        except errors.VersionError as e:
+            print("This client is incompatible: " + str(e))
+            return 1
+        except errors.NotFound:
+            logger.debug("No SMB service principal exists, OK to proceed")
+        except errors.PublicError as e:
+            print(
+                "Cannot connect to the server due to "
+                "a generic error: %s" % str(e)
+            )
+            return 1
+
+        # At this point we have proper setup:
+        # - we connected to IPA API end-point as a host principal
+        # - no cifs/... principal exists so we can create it
+        print("Chosen IPA master: %s" % server)
+        print("SMB principal to be created: %s" % api.env.smb_princ)
+        print("NetBIOS name to be used: %s" % options.netbiosname)
+        logger.info("Chosen IPA master: %s", server)
+        logger.info("SMB principal to be created: %s", api.env.smb_princ)
+        logger.info("NetBIOS name to be used: %s", options.netbiosname)
+
+        # 1. Pull down ID range and other details of known domains
+        domains = retrieve_domain_information(api)
+        str_info = pretty_print_domain_information(domains)
+        logger.info("Discovered domains to use:\n%s", str_info)
+        print("Discovered domains to use:\n%s" % str_info)
```

Comment:
What if no domains are returned? Should this be allowed to continue?

Reply:
At least our own domain should be discovered. I added the code that checks for two parts:
- `ipa trustconfig-show` actually available through the metadata
- `ipa trustconfig-show` returns at least information about our domain